### PR TITLE
IGNITE-23128 Use a placeholder StorageEngine for tables with unknown storage engine

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-23128

### CODE REVIEW

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/NullMvTableStorage.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * {@link MvTableStorage} implementation used when the node currently does not have the storage profile declared for the table
+ * and/or the corresponding storage engine. Such an object is used as a placeholder to have ability to create a table object,
+ * but any storage functionality will be absent (exceptions will be thrown).
+ *
+ * <p>Before a node starts getting assignments for partitions of such a table, both storage engine and storage profile must be created
+ * on the node and this placeholder object must be replaced with a 'real' storage object.
+ */
+public class NullMvTableStorage implements MvTableStorage {
```
Comment (ibessonov) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/NullMvTableStorage.java:38]: I would prefer not to have a table storage instance at all, but that is a much more involved change in the code. We should consider it in the future. Current approach feels more like a warkaround

Comment (ibessonov) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/NullMvTableStorage.java:38]: Should we move it into `table` module?

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/NullStorageEngine.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * {@link StorageEngine} implementation used when the node currently does not have the storage profile declared for the table
+ * and/or the corresponding storage engine. Such an object is used as a placeholder to have ability to create a table object,
+ * but any storage functionality will be absent (exceptions will be thrown).
+ *
+ * <p>Before a node starts getting assignments for partitions of such a table, both storage engine and storage profile must be created
+ * on the node and this placeholder object must be replaced with a 'real' storage object.
+ */
+public class NullStorageEngine implements StorageEngine {
```
Comment (ibessonov) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/NullStorageEngine.java:32]: Same comments here

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 3]
```diff
@@ -1737,7 +1738,10 @@ private CompletableFuture<List<Assignments>> getOrCreateAssignments(
     protected MvTableStorage createTableStorage(CatalogTableDescriptor tableDescriptor, CatalogZoneDescriptor zoneDescriptor) {
         StorageEngine engine = dataStorageMgr.engineByStorageProfile(tableDescriptor.storageProfile());
 
-        assert engine != null : "tableId=" + tableDescriptor.id() + ", storageProfile=" + tableDescriptor.storageProfile();
+        if (engine == null) {
```
Comment (ibessonov) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1741]: Does everything else work?

Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1741]: Could you please elaborate what `everything else` means here? At least, tests are green.

Comment (ibessonov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1741]: I mean, that's an arbitrary question. I'm a little shocked that it's one of just a few places that you had to modify

**Review by ibessonov — APPROVED** [review]

**Pull request merged**